### PR TITLE
[DRAFT] Orthogonal unicode graphs

### PR DESCRIPTION
**Pull request by MForster**
This is only intended for discussion, not for submission. I'd like to understand if you'd be interested in having an optional alternative commit graph formatting (or parts of it). Please don't look at the implementation, but only at the different insta test results, in particular in `graphlog.rs`. The code would need cleanup and proper factoring with the current ASCII-based implementation.

I've been playing around a bit, trying to improve (for my subjective view) the readability of the commit graphs. The basica ideas are these:
- Diagonal lines can be very hard to follow, in particular if there are many of them. This style uses only horizontal and vertical lines. This makes the graphs a bit wider and shorter.
- Using unicode box drawing characters makes the lines more continuous and easier to follow.

```
        ⬤ node 7
        │ ⬤ node 6
        │ │ ⬤ node 5
        │ │ │ ⬤ node 4
        ├─│─│─╯ 
        │ ⬤ │ node 3
        ├─╯ │ 
        │   ⬤ node 2
        ├───╯ 
        ⬤ node 1
```

image

I'll add more comments to the examples in the source files.

# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have added tests to cover my changes

### CODE REVIEW

File: src/graphlog.rs [hunk 1]
```diff
@@ -339,14 +298,14 @@ mod tests {
         let mut buffer = vec![];
         let mut graph = AsciiGraphDrawer::new(&mut buffer);
         graph.add_node(&2, &[Edge::direct(1)], b"@", b"many\nlines\nof\ntext\n")?;
-        graph.add_node(&1, &[], b"o", b"single line")?;
+        graph.add_node(&1, &[], "⬤".as_bytes(), b"single line")?;
```
Comment (MForster) [inline src/graphlog.rs:302]: My suggestion would be to make the actual characters configurable.

File: src/graphlog.rs [hunk 2]
```diff
@@ -558,30 +517,30 @@ mod tests {
...
-        o node 1
+        ⬤ node 10
+        │ ⬤ node 9
+        │ │ ⬤ node 8
+        ⬤ │ │ node 7
+        │ ⬤ │ node 6
+        │ │ ⬤ node 5
+        ⬤ │ │ node 4
+        │ ⬤ │ node 3
+        ├─╯ │
```
Comment (MForster) [inline src/graphlog.rs:540]: Here would be more potential for saving space. The blank line is not strictly necessary.

File: src/graphlog.rs [hunk 3]
```diff
@@ -707,37 +663,41 @@ mod tests {
...
-        |/  
-        o node 1
+        ⬤ node 8
+        │ ⬤ node 7
+        │ │ ⬤ node 6
+        │ │ │ with
+        │ │ │ some
+        │ │ │ more
+        │ │ │ lines
+        │ ⬤─│─╮ node 5
```
Comment (MForster) [inline src/graphlog.rs:693]: There can be more crossing. That's the price for straight edges together with a greedy algorithm...

File: src/graphlog.rs [hunk 4]
```diff
@@ -864,20 +820,20 @@ mod tests {
...
-        | o---.   node 2
-        | |\ \ \  with
-        | | : ~/  many
-        | ~/ /    lines
-        |/ /      of
-        |/        text
-        o node 1
+        ⬤ node 3
+        │ ⬤─┬─┬─╮ node 2
+        │ │ ┊ ⯆ ┊ with
```
Comment (MForster) [inline src/graphlog.rs:831]: Those who missing edges could end in the same line.

File: src/graphlog.rs [hunk 5]
```diff
@@ -864,20 +820,20 @@ mod tests {
...
+        │ ⯆ ┊   ┊ many
+        ├───╯   ┊ lines
+        ├───────╯ of
```
Comment (MForster) [inline src/graphlog.rs:834]: The horizontal parts of indirect edges should be dashed as well.

File: src/graphlog-comparison/sapling-curved.txt [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+        ├─╯ │
+        │   o  node 2
+        ├───╯
+        o  node 1
+        
+
+        o  node 5
+        │ o  node 4
+        │ │ o  node 3
+        ├───╯
```
Comment (MForster) [inline src/graphlog-comparison/sapling-curved.txt:114]: I think I like it better if the horizontal line is broken, not the vertical one (the horizontal gap is shorter)...

File: src/graphlog-comparison/sapling-curved.txt [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+        │   │     o  node 3
+        ├─────────╯
+        │   o  node 2
+        ├───╯
+        o  node 1
+        
+
+        @        node 1
+        ├─┬─┬─╮  with
+        │ │ │ │  many
```
Comment (MForster) [inline src/graphlog-comparison/sapling-curved.txt:211]: Not sure why this doesn't use dashed lines for the indirect edges. Might be a bug.

File: src/graphlog-comparison/sapling-curved.txt (hunk 6 above)
Comment (quark-zju) [reply src/graphlog-comparison/sapling-curved.txt:114]: cc @markbt for consideration.

File: src/graphlog-comparison/sapling-curved.txt (hunk 7 above)
Comment (quark-zju) [reply src/graphlog-comparison/sapling-curved.txt:211]: I think this is because `~` currently means "an anonymous direct parent and we stop here - skip rendering its ancestors". `~` does not mean "a (distant) ancestor or root". A `~` might be one single commit without ancestors, in which case dashed lines are not appropriate.

Implementation wise,  `~` is used by [`Ancestor::Anonymous`](https://github.com/facebook/sapling/blob/5093c59f3f2fea4a3a4953596ea2b30e4d50292a/eden/scm/lib/dag/src/render/render.rs#L53). Only `Ancestor::Ancestor(_)` uses dashed lines.

Comment (MForster) [reply src/graphlog-comparison/sapling-curved.txt:211]: Yes, there are supposed to be `Ancestor` edges. Unless I have a bug.

Comment (quark-zju) [reply src/graphlog-comparison/sapling-curved.txt:211]: That seems strange. I think `Ancestor::Ancestor(_)` nodes occupy one line per node so it can have associated  commit messages. IIUC it's only possible to have multiple `~`s in one line by using `Ancestor::Anonymous`.

Comment (MForster) [reply src/graphlog-comparison/sapling-curved.txt:211]: Sorry, I commented on the wrong graph. It's the one below ([line 200](https://github.com/martinvonz/jj/pull/1048/files/12e17d4b689ba3723c681c6f1f8cdf40f58c7c70#diff-a97eeb0ef7cc0f5332817fe118d7d910566d6cb857380b2b9853de17248d174fR200) vs. [line 220](https://github.com/martinvonz/jj/pull/1048/files/12e17d4b689ba3723c681c6f1f8cdf40f58c7c70#diff-1ca0bda506fe1e2ffd7e9bc08dd53dca9fcd80a923e97c4d97c946703aa2a56fR220)). I think the "ancestor" part may just have length 0 in that drawing...

But let me double-check first that I'm calling this incorrectly. I'll get back to you.
